src/unix/stream.c

```diff
@@ -839,11 +840,31 @@ static void uv__write(uv_stream_t* stream) {
 #endif
   } else {
     do {
+      /* On some platforms, notably macOS, the sum of iov_len must not
+       * overflow a 32-bit signed integer.
+       */
+      size_t old_iov_len = iov[iovcnt - 1].iov_len;
+      if (stream->write_queue_size > INT32_MAX) {
+        int32_t total_bytes = 0;
+        int new_iov_cnt;
+        for (new_iov_cnt = 0; new_iov_cnt < iovcnt; new_iov_cnt++) {
+          old_iov_len = iov[new_iov_cnt].iov_len;
+          if (((int64_t)total_bytes + iov[new_iov_cnt].iov_len) >= INT32_MAX) {
+            iov[new_iov_cnt].iov_len = INT32_MAX - total_bytes;
+            break;
+          }
+          total_bytes += iov[new_iov_cnt].iov_len;
+        }
+        iovcnt = new_iov_cnt + 1;
+       }
+
       if (iovcnt == 1) {
         n = write(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len);
       } else {
         n = writev(uv__stream_fd(stream), iov, iovcnt);
       }
+
+      iov[iovcnt - 1].iov_len = old_iov_len; /* restore original size of write in iov buf */
```

Comment:
Can I suggest a different approach that is IMO a little cleaner and simpler:
```diff
diff --git a/src/unix/stream.c b/src/unix/stream.c
index 6fc0a01f..fabf543b 100644
--- a/src/unix/stream.c
+++ b/src/unix/stream.c
@@ -744,7 +744,7 @@ static int uv__handle_fd(uv_handle_t* handle) {
 }
 
 static void uv__write(uv_stream_t* stream) {
-  struct iovec* iov;
+  uv_buf_t* iov;
   QUEUE* q;
   uv_write_t* req;
   int iovmax;
@@ -768,7 +768,7 @@ start:
    * because Windows's WSABUF is not an iovec.
    */
   assert(sizeof(uv_buf_t) == sizeof(struct iovec));
-  iov = (struct iovec*) &(req->bufs[req->write_index]);
+  iov = req->bufs + req->write_index;
   iovcnt = req->nbufs - req->write_index;
 
   iovmax = uv__getiovmax();
@@ -804,7 +804,7 @@ start:
 
     msg.msg_name = NULL;
     msg.msg_namelen = 0;
-    msg.msg_iov = iov;
+    msg.msg_iov = (struct iovec*) iov;
     msg.msg_iovlen = iovcnt;
     msg.msg_flags = 0;
 
@@ -838,11 +838,16 @@ start:
     while (n == -1 && errno == EINTR);
 #endif
   } else {
+    if (iovcnt > 1 && uv__count_bufs(iov, iovcnt) > INT32_MAX)
+      iovcnt = 1;
+
     do {
       if (iovcnt == 1) {
-        n = write(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len);
+        n = write(uv__stream_fd(stream),
+                  iov[0].base,
+                  iov[0].len > INT32_MAX ? INT32_MAX : iov[0].len);
       } else {
-        n = writev(uv__stream_fd(stream), iov, iovcnt);
+        n = writev(uv__stream_fd(stream), (struct iovec*) iov, iovcnt);
       }
     }
 #if defined(__APPLE__)
```
Aside: `iov` and `iovcnt` should probably be renamed to `bufs` and `nbufs` respectively.

(And also this:)
<details>

```diff
diff --git a/src/uv-common.c b/src/uv-common.c
index bc7d1379..b1cd5f48 100644
--- a/src/uv-common.c
+++ b/src/uv-common.c
@@ -431,13 +431,13 @@ uint64_t uv_now(const uv_loop_t* loop) {
 
 
 
-size_t uv__count_bufs(const uv_buf_t bufs[], unsigned int nbufs) {
+uint64_t uv__count_bufs(const uv_buf_t bufs[], unsigned int nbufs) {
   unsigned int i;
-  size_t bytes;
+  uint64_t bytes;
 
   bytes = 0;
   for (i = 0; i < nbufs; i++)
-    bytes += (size_t) bufs[i].len;
+    bytes += bufs[i].len;
 
   return bytes;
 }
diff --git a/src/uv-common.h b/src/uv-common.h
index 781a8559..f2fb1f03 100644
--- a/src/uv-common.h
+++ b/src/uv-common.h
@@ -120,7 +120,7 @@ void uv__work_submit(uv_loop_t* loop,
 
 void uv__work_done(uv_async_t* handle);
 
-size_t uv__count_bufs(const uv_buf_t bufs[], unsigned int nbufs);
+uint64_t uv__count_bufs(const uv_buf_t bufs[], unsigned int nbufs);
 
 int uv__socket_sockopt(uv_handle_t* handle, int optname, int* value);

```
</details>

Reply:
So when you `writev` more than `INT32_MAX` bytes total, if reverts to doing `write` instead? I'm not sure I fully understand the benefit of that. (but then, I haven't hooked up `writev` usage yet either in my application, so it doesn't yet matter to me).